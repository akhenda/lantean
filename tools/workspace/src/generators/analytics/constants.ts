/**
 * Default name of the library.
 */
export const defaultLibName = 'analytics';

/**
 * Default directory of the library.
 */
export const defaultLibDirectory = '';

/**
 * Default tags of the library.
 */
export const defaultLibTags = ['web', 'mobile', 'server', 'client', 'analytics'];

/**
 * Dependencies of the library.
 */
export const deps = {
  'expo-application': '^5.9.1',
  'expo-device': '^6.0.2',
  'expo-file-system': '^17.0.1',
  'expo-localization': '^15.0.3',
  'posthog-js': '^1.181.0',
  'posthog-node': '^4.2.1',
  'posthog-react-native': '^3.3.10',
} as const;
