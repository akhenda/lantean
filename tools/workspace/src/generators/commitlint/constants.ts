import { CommitlintConfigFileName } from './types';

export const commitlintConfigFile: CommitlintConfigFileName =
  '.commitlintrc.js';

export const devDependencies = {
  '@commitlint/cli': '^19.4.0',
  '@commitlint/config-conventional': '^19.2.2',
  '@commitlint/config-nx-scopes': '^19.3.1',
  '@theunderscorer/nx-semantic-release': '^2.12.0',
  'cz-git': '^1.9.4',
} as const;
