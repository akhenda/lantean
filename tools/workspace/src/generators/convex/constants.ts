/**
 * Default name of the library.
 */
export const defaultLibName = 'backend';

/**
 * Default directory of the library.
 */
export const defaultLibDirectory = '';

/**
 * Default tags of the library.
 */
export const defaultLibTags = ['server', 'backend', 'type:convex', 'domain:server'];

/**
 * Dependencies of the library.
 */
export const deps = {
  '@edge-runtime/vm': '^4.0.4',
  convex: '^1.17.0',
  'convex-test': '^0.0.34',
} as const;

export const devDeps = {} as const;
