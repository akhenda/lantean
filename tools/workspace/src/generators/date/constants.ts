/**
 * Default name of the library.
 */
export const defaultLibName = 'date';

/**
 * Default directory of the library.
 */
export const defaultLibDirectory = '';

/**
 * Default tags of the library.
 */
export const defaultLibTags = ['web', 'mobile', 'server', 'client', 'date'];

/**
 * Dependencies of the library.
 */
export const deps = {
  'dayjs': '^1.11.13',
} as const;
