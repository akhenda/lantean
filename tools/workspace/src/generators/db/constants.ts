/**
 * Default name of the library.
 */
export const defaultLibName = 'db';

/**
 * Default directory of the library.
 */
export const defaultLibDirectory = '';

/**
 * Default tags of the library.
 */
export const defaultLibTags = ['server', 'db'];

/**
 * Dependencies of the library.
 */
export const deps = {
  '@clerk/nextjs': '^6.3.2',
  '@clerk/themes': '^2.1.43',
  dotenv: '^16.4.5',
  'dotenv-expand': '^11.0.6',
  'drizzle-orm': '^0.36.1',
  'drizzle-zod': '^0.5.1',
  pg: '^8.13.1',
  postgres: '^3.4.5',
  ulid: '^2.3.0',
  zod: '^3.23.8',
} as const;

export const devDeps = {
  'cross-env': '^7.0.3',
  'drizzle-kit': '^0.28.0',
} as const;
