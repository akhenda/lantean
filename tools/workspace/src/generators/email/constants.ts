/**
 * Default name of the library.
 */
export const defaultLibName = 'email';

/**
 * Default directory of the library.
 */
export const defaultLibDirectory = '';

/**
 * Default tags of the library.
 */
export const defaultLibTags = ['server', 'email'];

/**
 * Dependencies of the library.
 */
export const deps = {
  '@react-email/components': '^0.0.27',
  '@react-email/tailwind': '^1.0.1',
} as const;
