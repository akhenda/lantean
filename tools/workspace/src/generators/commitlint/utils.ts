import { Tree } from '@nx/devkit';

import { NormalizedSchema } from './types';
import { CommitlintGeneratorSchema } from './schema';
import { defaultCommitlintConfigFile } from './constants';

import { getGitRepo } from '../../devkit';

/**
 * Normalize options for the commitlint generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `fileName`: The path to the commitlint configuration file. If not provided, the default value is {@link defaultCommitlintConfigFile}.
 * - `gitRepo`: The URL of the git repository.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 */
export function normalizeOptions(
  tree: Tree,
  options: CommitlintGeneratorSchema,
): NormalizedSchema {
  const fileName = options ? options.configFileName : defaultCommitlintConfigFile;

  return {
    ...options,
    fileName,
    gitRepo: getGitRepo(tree)?.replace(/^git@github\.com:/, 'https://github.com/'),
  };
}

/**
 * The content of the workspace README.md file.
 */
export const readmeContent = `
## Docs
[Commitlint & Semantic Release](./docs/commitlint.md)

## Run tasks
`;
