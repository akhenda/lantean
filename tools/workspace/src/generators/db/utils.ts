import { getWorkspaceLayout, names, offsetFromRoot, Tree } from '@nx/devkit';

import { defaultLibDirectory, defaultLibName, defaultLibTags } from './constants';
import { NormalizedSchema, DbGeneratorSchema } from './schema';

import { NormalizedSchema as LoggingLibNormalizedSchema } from '../logging/schema';
import { getImportPath, getNpmScope } from '../../utils';

/**
 * Normalize options for the DB generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @param extra The extra options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `name`: The name of the library.
 * - `projectName`: The name of the project.
 * - `projectRoot`: The path to the project root.
 * - `projectDirectory`: The path to the project directory.
 * - `importPath`: The import path for the library.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `parsedTags`: The parsed tags for the project.
 * - `npmScope`: The npm scope for the library.
 * - `npmScopeTitle`: The title cased npm scope for the library.
 * - `offsetFromRoot`: The relative path from the workspace root to the library.
 * - `loggingLibName`: The name of the logging library.
 * - `loggingLibImportPath`: The import path for the logging library.
 * - `typesLibName`: The name of the types library.
 * - `typesLibImportPath`: The import path for the types library.
 */
export function normalizeOptions(
  tree: Tree,
  options: DbGeneratorSchema,
  extra: Pick<LoggingLibNormalizedSchema, 'projectName' | 'importPath' | 'typesLibName' | 'typesLibImportPath'>,
): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const name = names(options.name ?? defaultLibName).fileName;
  const project = names(defaultLibDirectory).fileName;
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;
  const projectDirectory = `${project}/${name}`;
  const projectRoot = `${libsDir}/${projectDirectory}`;
  const importPath = getImportPath(tree, name);
  const npmScope = getNpmScope(tree) ?? name;
  const npmScopeTitle = names(npmScope).className;
  const offsetFromRootPath = offsetFromRoot(projectRoot);

  return {
    ...options,
    appsDir,
    importPath,
    libsDir,
    name: options.name,
    parsedTags: defaultLibTags,
    projectDirectory,
    projectName: name,
    projectRoot,
    npmScope,
    npmScopeTitle,
    offsetFromRoot: offsetFromRootPath,
    loggingLibName: extra.projectName,
    loggingLibImportPath: extra.importPath,
    typesLibName: extra.typesLibName,
    typesLibImportPath: extra.typesLibImportPath,
  };
}
