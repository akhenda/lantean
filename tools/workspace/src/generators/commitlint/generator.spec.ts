import {
  addDependenciesToPackageJson,
  installPackagesTask,
  readJson,
  Tree,
  writeJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import { addHuskyHook, getGitRepo } from '../../devkit';
import * as utils from '../../utils';

import { devDependencies } from './constants';
import commitlintGenerator from './generator';

jest.mock('../../devkit', () => ({
  ...jest.requireActual('../../devkit'),
  addHuskyHook: jest.fn(),
  getGitRepo: jest.fn(),
}));

jest.mock('@nx/devkit', () => ({
  ...jest.requireActual('@nx/devkit'),
  installPackagesTask: jest.fn(),
  addDependenciesToPackageJson: jest.fn(),
}));

const mockGetGitRepo = getGitRepo as jest.Mock;

describe('@lantean/workspace commitlint generator', () => {
  let tree: Tree;
  const gitRepo = 'https://github.com/akhenda/lantean';

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    jest.spyOn(console, 'log').mockImplementation(() => null);
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it('should run successfully', async () => {
    await commitlintGenerator(tree);
  });

  it('should add dependencies', async () => {
    await commitlintGenerator(tree);

    expect(addDependenciesToPackageJson).toHaveBeenCalledWith(
      tree,
      {},
      devDependencies
    );
  });

  it('should add husky hook', async () => {
    jest.spyOn(utils, 'isHuskyInstalled').mockReturnValueOnce(true);

    await commitlintGenerator(tree);

    expect(addHuskyHook).toHaveBeenCalledWith(
      tree,
      'commit-msg',
      'npx --no-install commitlint --edit $1'
    );
  });

  it('should create commitlint config', async () => {
    const configFileName = '.commitlintrc';
    await commitlintGenerator(tree, { configFileName });

    expect(readJson(tree, configFileName)).toMatchSnapshot();
  });

  it('should merge commitlint config', async () => {
    const configFileName = '.commitlintrc.json';
    const commitlintEmpty = { extends: [], rules: {} };

    mockGetGitRepo.mockReturnValue(gitRepo);

    writeJson(tree, configFileName, commitlintEmpty);

    await commitlintGenerator(tree, { configFileName });

    expect(readJson(tree, configFileName)).toMatchSnapshot();
  });

  it('should run tasks', async () => {
    jest.spyOn(utils, 'isHuskyInstalled').mockReturnValueOnce(false);

    const tasks = await commitlintGenerator(tree);

    expect(tasks).toBeTruthy();

    tasks?.();

    expect(utils.isHuskyInstalled).toHaveBeenCalled();
    expect(installPackagesTask).toHaveBeenCalled();
    expect(addHuskyHook).not.toHaveBeenCalled();
  });
});
