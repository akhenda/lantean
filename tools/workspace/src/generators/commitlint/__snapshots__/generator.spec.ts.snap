// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@lantean/workspace commitlint generator should create commitlint config 1`] = `
{
  "extends": [
    "@commitlint/config-conventional",
    "@commitlint/config-nx-scopes",
  ],
  "prompt": {
    "aiNumber": 3,
    "aiQuestionCB": "Write an insightful and concise Git commit message in the present tense for the following Git diff code, without any prefixes, and no longer than 72 characters",
    "alias": {
      "fd": "docs: fix typos",
    },
    "allowBreakingChanges": [
      "feat",
      "fix",
    ],
    "allowCustomIssuePrefix": true,
    "allowCustomScopes": true,
    "allowEmptyIssuePrefix": true,
    "allowEmptyScopes": true,
    "breaklineChar": "|",
    "breaklineNumber": 100,
    "confirmColorize": true,
    "customIssuePrefixAlias": "custom",
    "customIssuePrefixAlign": "top",
    "customScopesAlias": "custom",
    "customScopesAlign": "bottom",
    "defaultBody": "",
    "defaultIssues": "",
    "defaultScope": "",
    "defaultSubject": "",
    "emojiAlign": "center",
    "emptyIssuePrefixAlias": "skip",
    "emptyScopesAlias": "empty",
    "issuePrefixes": [
      {
        "name": "closed:   ISSUES has been processed",
        "value": "closed",
      },
    ],
    "markBreakingChangeMode": false,
    "messages": {
      "body": "Provide a LONGER description of the change (optional). Use "|" to break new line:
",
      "breaking": "List any BREAKING CHANGES (optional). Use "|" to break new line:
",
      "confirmCommit": "Are you sure you want to proceed with the commit above?",
      "customFooterPrefix": "Input ISSUES prefix:",
      "customScope": "Denote the SCOPE of this change:",
      "footer": "List any ISSUES by this change. E.g.: #31, #34:
",
      "footerPrefixesSelect": "Select the ISSUES type of changeList by this change (optional):",
      "generatedSelectByAI": "Select suitable subject by AI generated:",
      "generatingByAI": "Generating your AI commit subject...",
      "scope": "Denote the SCOPE of this change (optional):",
      "subject": "Write a SHORT, IMPERATIVE tense description of the change:
",
      "type": "Select the type of change that you're committing:",
    },
    "minSubjectLength": 0,
    "skipQuestions": [],
    "themeColorCode": "",
    "types": [
      {
        "emoji": ":sparkles:",
        "name": "feat:     ‚ú®  A new feature",
        "value": "feat",
      },
      {
        "emoji": ":bug:",
        "name": "fix:      üêõ  A bug fix",
        "value": "fix",
      },
      {
        "emoji": ":recycle:",
        "name": "refactor: ‚ôªÔ∏è   A code change that neither fixes a bug nor adds a feature",
        "value": "refactor",
      },
      {
        "emoji": ":package:",
        "name": "build:    üì¶Ô∏è   Changes that affect the build system or external dependencies",
        "value": "build",
      },
      {
        "emoji": ":hammer:",
        "name": "chore:    üî®  Other changes that don't modify src or test files",
        "value": "chore",
      },
      {
        "emoji": ":memo:",
        "name": "docs:     üìù  Documentation only changes",
        "value": "docs",
      },
      {
        "emoji": ":lipstick:",
        "name": "style:    üíÑ  Markup, white-space, formatting, missing semi-colons...",
        "value": "style",
      },
      {
        "emoji": ":zap:",
        "name": "perf:     ‚ö°Ô∏è  A code change that improves performance",
        "value": "perf",
      },
      {
        "emoji": ":white_check_mark:",
        "name": "test:     ‚úÖ  Adding missing tests or correcting existing tests",
        "value": "test",
      },
      {
        "emoji": ":ferris_wheel:",
        "name": "ci:       üé°  Changes to our CI configuration files and scripts",
        "value": "ci",
      },
      {
        "emoji": ":rewind:",
        "name": "revert:   ‚è™Ô∏è  Reverts a previous commit",
        "value": "revert",
      },
    ],
    "upperCaseSubject": false,
    "useAI": false,
    "useEmoji": true,
  },
  "rules": {
    "scope-enum": [
      "deps",
      "version",
      "release",
      "deploy",
      "configs",
      "ds",
    ],
  },
}
`;

exports[`@lantean/workspace commitlint generator should merge commitlint config 1`] = `
{
  "extends": [],
  "rules": {},
}
`;
